Common/Orders/Fills/EquityFillModel.cs

```diff
@@ -13,204 +13,174 @@
+        /// So we have to assume a fluid, high volume market, where every price between the high and low has been touched.</returns>
         /// <seealso cref="MarketFill(Security, MarketOrder)"/>
-        public virtual OrderEvent StopMarketFill(Security asset, StopMarketOrder order)
+        public override OrderEvent StopMarketFill(Security asset, StopMarketOrder order)
         {
-            //Default order event to return.
+            // Default order event to return.
             var utcTime = asset.LocalTime.ConvertToUtc(asset.Exchange.TimeZone);
             var fill = new OrderEvent(order, utcTime, OrderFee.Zero);
 
-            //If its cancelled don't need anymore checks:
-            if (order.Status == OrderStatus.Canceled) return fill;
-
-            // make sure the exchange is open/normal market hours before filling
-            if (!IsExchangeOpen(asset, false)) return fill;
+            if (order.Status == OrderStatus.Canceled ||            // Orders with canceled status or hold
+                order.Direction == OrderDirection.Hold ||          // direction don't need anymore checks
+                !IsExchangeOpen(asset, false))   // Exchange need to be open/normal market hours
+            {
+                return fill;
+            }
 
-            //Get the range of prices in the last bar:
-            var prices = GetPricesCheckingPythonWrapper(asset, order.Direction);
-            var pricesEndTime = prices.EndTime.ConvertToUtc(asset.Exchange.TimeZone);
+            // Get the last trade bar since stop orders are triggered by trades
+            var lastTradeBar = GetLastTradeBar(asset);
 
-            // do not fill on stale data
-            if (pricesEndTime <= order.Time) return fill;
+            // Do not fill on stale data
+            var localOrderTime = order.Time.ConvertFromUtc(asset.Exchange.TimeZone);
+            if (lastTradeBar.EndTime <= localOrderTime) return fill;
 
-            //Calculate the model slippage: e.g. 0.01c
+            // Calculate the model slippage: e.g. 0.01c
             var slip = asset.SlippageModel.GetSlippageApproximation(asset, order);
 
-            //Check if the Stop Order was filled: opposite to a limit order
+            // Check if the Stop Order was filled: opposite to a limit order
             switch (order.Direction)
             {
-                case OrderDirection.Sell:
-                    //-> 1.1 Sell Stop: If Price below setpoint, Sell:
-                    if (prices.Low < order.StopPrice)
+                case OrderDirection.Buy:
+                    // Buy Stop: If Price Above set point, Buy:
+                    if (lastTradeBar.High >= order.StopPrice)
                     {
+                        // Bar opens above stop price, fill at open price
+                        if (lastTradeBar.Open >= order.StopPrice)
+                        {
+                            fill.FillPrice = lastTradeBar.Open + slip;
```

Comment:
This one is another nice one to add to the PR description, we might give it a review see what other behavior change might be missing